socorro/lib/libjsonschema.py

```diff
@@ -100,39 +188,29 @@ def convert_to(document_part, target_format):
     raise UnknownConvertFormat(f"{target_format!r} is an unknown format")
 
 
-# Cache of pattern -> re object
-PATTERN_CACHE = {}
-
-
-def compile_pattern_re(pattern):
-    """Compile a patternProperties regex pattern
+class JsonSchemaReducer:
+    """Reducer for reducing a document to the structure of the specified jsonschema
 
-    Note that uses a module-level cache to reduce repeated pattern compiling.
+    Several things to know about the reducer:
 
-    :arg string pattern: a patternProperties regex pattern
-
-    :returns: re.Pattern
+    1. It doesn't support full jsonschema spec--it only supports the parts we
+       needed for our schemas.
+    2. It does some light type validation and raises exceptions for documents that
+       are invalid.
 
     """
-    if pattern not in PATTERN_CACHE:
-        PATTERN_CACHE[pattern] = re.compile(pattern)
-    return PATTERN_CACHE[pattern]
-
 
-class Reducer:
-    def __init__(self, schema, include_predicate=everything_predicate):
```

Comment:
This kind of reducer was interesting, but in practice it was difficult to use because it was doing too many things at the same time. In the "new plan", I broke this up into three things:

1. resolve references producing a resolved schema
2. transform a schema into another schema using some transform function
3. reduce a document using a schema

That's much easier to work with and reason about and made the reducer code more straight-forward.